ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala

```diff
@@ -18,118 +18,164 @@
 package com.spotify.ratatool.diffy
 
 import com.spotify.scio.coders.Coder
-import org.apache.avro.{Schema, SchemaValidatorBuilder}
-import org.apache.avro.generic.GenericRecord
-import scala.util.Try
+import org.apache.avro.{Conversion, Schema}
+import org.apache.avro.generic.{GenericData, GenericRecord, IndexedRecord}
+import org.apache.avro.specific.SpecificData
 
 import scala.jdk.CollectionConverters._
+import scala.util.Try
 
 /** Field level diff tool for Avro records. */
-class AvroDiffy[T <: GenericRecord: Coder](
+class AvroDiffy[T <: IndexedRecord: Coder](
   ignore: Set[String] = Set.empty,
   unordered: Set[String] = Set.empty,
   unorderedFieldKeys: Map[String, String] = Map()
 ) extends Diffy[T](ignore, unordered, unorderedFieldKeys) {
 
-  override def apply(x: T, y: T): Seq[Delta] = {
-    new SchemaValidatorBuilder().canReadStrategy
-      .validateLatest()
-      .validate(y.getSchema, List(x.getSchema).asJava)
```

Comment:
In terms of diff, avro schemas must be strictly equal. This is checked line 62 in the new version.

Compatible schema are used on read time to adapt stored data to the desired read schema. Once in memory, we should not compare data constructed with different schema, even if compatible.

Reply:
Can you expand on what you mean? I'm a bit confused reading this thread

Reply:
avro record with different models must not be compared.

schema compatibility is relevant when reading, making sure the `writerSchema` and the `readerSchema` are compatible. Once read, the records strictly folow the `readerSchema` where **field index** matters. Strict schema equality must be ensured before comparing content.

Reply:
Ok, trying to confirm my understanding here since I'm still a bit confused.
Dataset A is updated to add new nullable field `x` and becomes Dataset A'.
We go to diff these two datasets.
Are you saying that this field comparison will end up in one of the above null cases prior to the schema comparison?

Reply:
If the new field has a default value, we should compare the 2 datasets using either A or A' schema when reading.
- using the A schema -> the new field won't be present in both side when reading, diff won't find any issue.
- using the A' schema -> we'll get the diffs where the value is present in the new dataset.
Obtained schema for those records are always equal

Now, if we read A with A schema and A' with A' schema, we should not even compare records.

Reply:
I would go one step further and say it should necessarily be the A' schema, even if it's nullable/has default, and cases where a field is missing should fail. IIRC this is the current functionality. It's unclear to me if/where this behaviour is retained

Reply:
The [BigDiffy](https://github.com/spotify/ratatool/blob/51ec7981aa283b6ca07742474166666780afbcdc/ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala#L145) API of this lib is not file aware. It only works in terms of in-memory records and can't make any assumption on writer schema.

It is up to the users creating the `SCollection` to read using the correct schema

Reply:
> The [BigDiffy](https://github.com/spotify/ratatool/blob/51ec7981aa283b6ca07742474166666780afbcdc/ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala#L145) API of this lib is not file aware.

cc @RustedBones, AFAIK, BigDiffy is file-aware when run through the CLI, which invokes BigDiffy#run: https://github.com/spotify/ratatool/blob/51ec7981aa283b6ca07742474166666780afbcdc/ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala#L683-L760

but even then, it looks like it selects the [schema associated with the RHS](https://github.com/spotify/ratatool/blob/51ec7981aa283b6ca07742474166666780afbcdc/ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala#L751-L754) and uses that for both resulting SCollections. So maybe we could add schema validation there (ensure that RHS schema is equal to, or a superset of, the LHS schema)?

Reply:
Added an extra check that will prefer backward compatible reader schema.
When schemas are different, but both forward and backward compatible, will print a warning